modules/caddytls/capools.go

```diff
@@ -0,0 +1,803 @@
+func (sp *StoragePool) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
+	d.Next()
+	// if args are given
+	sp.PEMKeys = append(sp.PEMKeys, d.RemainingArgs()...)
+	for nesting := d.Nesting(); d.NextBlock(nesting); {
+		switch d.Val() {
+		case "storage":
+			if sp.StorageRaw != nil {
+				return d.Err("storage module already set")
+			}
+			if !d.Next() {
+				return d.ArgErr()
+			}
+			modID := "caddy.storage." + d.Val()
+			unm, err := caddyfile.UnmarshalModule(d, modID)
+			if err != nil {
+				return err
+			}
+			storage, ok := unm.(caddy.StorageConverter)
+			if !ok {
+				return d.Errf("module %s is not a caddy.StorageConverter", modID)
+			}
+			sp.StorageRaw = caddyconfig.JSONModuleObject(storage, "module", d.Val(), nil)
+		case "keys":
+			sp.PEMKeys = append(sp.PEMKeys, d.RemainingArgs()...)
+		default:
+			return fmt.Errorf("unrecognized directive: %s", d.Val())
+		}
+	}
+	return nil
+}
+
+func (p StoragePool) CertPool() *x509.CertPool {
+	return p.pool
+}
+
+// TLSConfig holds configuration related to the TLS configuration for the
+// transport/client.
+// copied from with minor modifications: modules/caddyhttp/reverseproxy/httptransport.go
+type TLSConfig struct {
+	// Provides the guest module that provides the trusted certificate authority (CA) certificates
+	CARaw json.RawMessage `json:"ca,omitempty" caddy:"namespace=tls.ca_pool.source inline_key=provider"`
+
+	// If true, TLS verification of server certificates will be disabled.
+	// This is insecure and may be removed in the future. Do not use this
+	// option except in testing or local development environments.
+	InsecureSkipVerify bool `json:"insecure_skip_verify,omitempty"`
+
+	// The duration to allow a TLS handshake to a server. Default: No timeout.
+	HandshakeTimeout caddy.Duration `json:"handshake_timeout,omitempty"`
+
+	// The server name used when verifying the certificate received in the TLS
+	// handshake. By default, this will use the upstream address' host part.
+	// You only need to override this if your upstream address does not match the
+	// certificate the upstream is likely to use. For example if the upstream
+	// address is an IP address, then you would need to configure this to the
+	// hostname being served by the upstream server. Currently, this does not
+	// support placeholders because the TLS config is not provisioned on each
+	// connection, so a static value must be used.
+	ServerName string `json:"server_name,omitempty"`
+
+	// TLS renegotiation level. TLS renegotiation is the act of performing
+	// subsequent handshakes on a connection after the first.
+	// The level can be:
+	//  - "never": (the default) disables renegotiation.
+	//  - "once": allows a remote server to request renegotiation once per connection.
+	//  - "freely": allows a remote server to repeatedly request renegotiation.
+	Renegotiation string `json:"renegotiation,omitempty"`
+}
+
+func (t *TLSConfig) unmarshalCaddyfile(d *caddyfile.Dispenser) error {
```

Comment:
If it's unexported then it doesn't conform to the interface. Maybe I'm wrong about how you mean to use this though.

```suggestion
func (t *TLSConfig) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
```

Reply:
This is intentional. I don't want to implement the interface 🙂 just for an internal usage and the method name is handy

Reply:
I think typically we use `parseConfig` or something instead, if it's not meant to be that interface

Reply:
but I think the intent here is clearer and requires less inference

Reply:
:man_shrugging: ok, I don't feel that strongly